src/docs/sphinx/dev_guide/weak_form.rst

```diff
@@ -0,0 +1,264 @@
+   }, mesh);
+
+   residual.AddSurfaceIntegral([](auto x, auto disp /* unused */){ return traction(x); }, surface_mesh);
+
+   auto r = residual(displacements);
+
+So, in only a few lines of code, we can create optimized, custom finite
+element kernels!
+
+.. _header-n34:
+
+Implementation
+==============
+
+For the most part, the ``WeakForm`` class is just a container of
+``Integral`` objects, and some prolongation and restriction operators to
+get the data they need:
+
+.. code-block:: cpp
+
+   template <typename test, typename trial>
+   struct WeakForm<test(trial)> : public mfem::Operator {
+     ...
+     std::vector< Integral<test(trial)> > domain_integrals;
+     std::vector< Integral<test(trial)> > boundary_integrals;
+   };
+
+The calls to ``WeakForm::Add****Integral`` forward the integrand and
+mesh information to an ``Integral`` constructor and add it to the
+appropriate list (either ``domain_integrals`` or
+``boundary_integrals``). MFEM treats domain and boundary integrals
+differently, so we maintain them in separate lists.
+
+From there, the ``Integral`` constructor uses the integrand functor to
+specialize a highly templated finite element kernel (simplified
+implementation given below).
+
+.. code-block:: cpp
+
+   template < ::Geometry g, typename test, typename trial, int geometry_dim, int spatial_dim, int Q,
+              typename derivatives_type, typename lambda>
+   void evaluation_kernel(const mfem::Vector& U, mfem::Vector& R, derivatives_type* derivatives_ptr,
+                          const mfem::Vector& J_, const mfem::Vector& X_, int num_elements, lambda qf)
+   {
+     ...
+
+     // for each element in the domain
+     for (int e = 0; e < num_elements; e++) {
+     
+       // get the values for this particular element
+       tensor u_elem = impl::Load<trial_element>(u, e);
+
+       // this is where we will accumulate the element residual tensor
+       element_residual_type r_elem{};
+
+       // for each quadrature point in the element
+       for (int q = 0; q < static_cast<int>(rule.size()); q++) {
+         // get the position of this quadrature point in the parent and physical space,
+         // and calculate the measure of that point in physical space.
+         auto   xi  = rule.points[q];
+         auto   dxi = rule.weights[q];
+         auto   x_q = make_tensor<spatial_dim>([&](int i) { return X(q, i, e); });
+         auto   J_q = make_tensor<spatial_dim, geometry_dim>([&](int i, int j) { return J(q, i, j, e); });
+         double dx  = impl::Measure(J_q) * dxi;
+
+         // evaluate the value/derivatives needed for the q-function at this quadrature point
+         auto arg = impl::Preprocess<trial_element>(u_elem, xi, J_q);
+
+         // evaluate the user-specified constitutive model
+         //
+         // note: make_dual(arg) promotes those arguments to dual number types
+         // so that qf_output will contain values and derivatives
+         auto qf_output = qf(x_q, make_dual(arg));
+
```

Comment:
I think it would be useful to have a section here on how to interpret the results of `get_gradient(dq_output)`. It's possible a user may want to get access to the gradient of a q function without performing an integration?

I think something to remind users about the `std::tuple<2,2>` nature of the output and what it represents would be sufficient.